extensions/pasteboard/watcher.m

```diff
@@ -0,0 +1,316 @@
+
+    // Trigger Lua Callback Function:
+    LuaSkin *skin = [LuaSkin sharedWithState:NULL];
+    _lua_stackguard_entry(skin.L);
+    
+    [skin pushLuaRef:refTable ref:self.fnRef];
+        
+    NSString *result = [pb stringForType:NSPasteboardTypeString];
+    if (result) {
+        [skin pushNSObject:result];
+    } else {
+        lua_pushnil(skin.L);
+    }
+
+    if (![skin protectedCallAndTraceback:1 nresults:0]) {
+        const char *errorMsg = lua_tostring(skin.L, -1);
+        [skin logBreadcrumb:[NSString stringWithFormat:@"hs.pasteboard.watcher callback error: %s", errorMsg]];
+        [skin logError:[NSString stringWithFormat:@"hs.pasteboard.watcher callback error: %s", errorMsg]];
+        lua_pop(skin.L, 1); // clear error message from stack
+    }
+    
+    _lua_stackguard_exit(skin.L);
+}
+
+- (void)start {
+    // Abort if the watcher is already running:
+    if (self.isRunning) {
+        return;
+    }
+    
+    // If the Shared Pasteboard Timer doesn't exist, create it:
+    if (!sharedPasteboardTimer.isValid) {
+        sharedPasteboardTimer = [NSTimer timerWithTimeInterval:POLLING_INTERVAL target:self selector:@selector(sharedPasteboardTimerCallback:) userInfo:nil repeats:YES];
+    }
+    
+    // Update Initial Change Count:
+    NSPasteboard *pb;
+    if (self.pbName) {
+        pb = [NSPasteboard pasteboardWithName:self.pbName];
+    } else {
+        pb = [NSPasteboard generalPasteboard];
+    }
+    self.changeCount = [pb changeCount];
+        
+    // Start the Shared Pasteboard NSTimer if it's not already running:
+    if (!CFRunLoopContainsTimer(CFRunLoopGetCurrent(), (__bridge CFRunLoopTimerRef)sharedPasteboardTimer, kCFRunLoopDefaultMode)) {
+        [[NSRunLoop currentRunLoop] addTimer:sharedPasteboardTimer forMode:NSRunLoopCommonModes];
+    }
+    
+    // Increment the General Pasteboard Timer Counter:
+    sharedPasteboardTimerCount++;
+    
+    // Add observer:
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(sharedPasteboardChanged:)
+                                                 name:@"sharedPasteboardNotification" object:nil];
+
+    // The watcher is now running:
+    self.isRunning = YES;
+}
+
+- (void)stop {
+    // Remove observer:
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:@"sharedPasteboardNotification"
+                                                  object:nil];
+
+    // Decrement the Shared Pasteboard Timer Counter:
+    sharedPasteboardTimerCount--;
+            
+    // If no more watchers are left, destroy the NSTimer:
+    if (sharedPasteboardTimerCount == 0) {
+        [sharedPasteboardTimer invalidate];
+        sharedPasteboardTimer = nil;
+    }
+    
+    // Watcher is no longer running:
+    self.isRunning = NO;
+}
+@end
+
+HSPasteboardTimer *createHSPasteboardTimer(int callbackRef, NSString *pbName) {
```

Comment:
Point of style: most often a subclass will have an initiallizer that does this kind of initial setup, rather than a helper function like this... something like the following up in the HSPasteboardTimer implementation:

~~~objc
+ (instancetype)initForName:(NSString *)pbName withCallback:(int)callbackRef {
    self = [super init] ;
    if (self) {
        _fnRef = callbackRef ;
        _pbname = pbName ;
        // or, assuming you haven't made the properties readonly, and prefer
        // the style, you could instead use:
        //     self.fnRef = callbackRef ;
        //     self.pbname = pbName ;
    }
    return self ;
}
~~~

and then in `hs.pasteboard.watcher.new`, use:

~~~objc
HSPasteboardTimer *timer = [[HSPasteboardTimer alloc] initForName:pbName withCallback:callbackRef] ;
~~~

but this works as well, so... your call.

Reply:
Good to know! I'm happy to try and keep things in-line with the rest of the Hammerspoon code, so more than happy to go with whatever you feel fits best.